### ISSUE TITLE
[HOLD for payment 2024-08-09] [HOLD for payment 2024-08-01] [$250] CRITICAL: Native - Message stays in the Composer after being sent

### ISSUE THREAD

**Issue by kacper-mikolajczak** [Expensify/App#37896]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.47-10
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** N/A
**Expensify/Expensify Issue URL:**
**Issue reported by:** @kacper-mikolajczak 
**Slack conversation:** https://callstack-hq.slack.com/archives/C05LX9D6E07/p1709816920314349

## Action Performed:

1. Open a report
2. Type and send messages one after another in fast manner

## Expected Result:

`Composer` is emptied of old message after `Send` button is pressed, resulting in clean state whenever user wants to type new message.

## Actual Result:

From time to time, `Composer` lags behind and old message is kept in the input.

## Related issues:

https://github.com/Expensify/App/issues/10148

https://github.com/Expensify/App/issues/35225#issuecomment-1912041169

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/62747088/e2ac7dfb-0302-4103-a3d1-e4272faf23c3

https://github.com/Expensify/App/assets/62747088/43d33c3d-8a79-4625-a168-90d29528e496

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010ec0505687d1ab91</li>
        <li>Upwork Job ID: 1807817457409282324</li>
        <li>Last Price Increase: 2024-07-01</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by kacper-mikolajczak** [comment]
Replacing `MarkdownInput` with regular `TextInput` did not help. 

https://github.com/Expensify/App/assets/62747088/5adc4164-8eb6-45b1-92a9-f4252e21b4f4

**Comment by muttmuure** [comment]
@kacper-mikolajczak is this still being worked on?

**Comment by kacper-mikolajczak** [comment]
Hi @muttmuure, sorry I am occupied with other tasks! Can we wait with final decision whether close the issue till Friday? Thanks a lot!

**Comment by kacper-mikolajczak** [comment]
Hi @muttmuure I want to let you know that I've been mostly involved in Monitoring setup and finishing the [`hasDraft` PR](https://github.com/Expensify/App/pull/37281) and was not able to get back to this assignment. 

I will continue with monitoring setup (and Control phase in general) next week, so let me know if you want to close or put on hold this issue, so we can get back to it later. Thanks a lot and have a nice peaceful weekend!

**Comment by muttmuure** [comment]
Let's put it on hold

**Comment by muttmuure** [comment]
Still held

**Comment by muttmuure** [comment]
Held

**Comment by muttmuure** [comment]
Held

**Comment by kevinksullivan** [comment]
asking about this [here](https://expensify.slack.com/archives/C05LX9D6E07/p1717540319284149)

**Comment by muttmuure** [comment]
Taking this off HOLD

**Comment by kacper-mikolajczak** [comment]
Hi folks! 

After investigating the Composer setup and getting familiar with [original discussion](https://github.com/Expensify/App/issues/10148), I am afraid that issue originates from how RN handles updates between JS and UI thread in general. 

The issue remains even with minimal reproduction sample where we only have input and the button that is suppose to clear it. Same was done previously [here](https://github.com/facebook/react-native/issues/37722), but there was no further discussion on how to avoid such behaviour. 

I will try to set up new arch repro environment to see if bridge-less communication is going to help with this inherent issue by comparing it to previous behaviour.

**Comment by kacper-mikolajczak** [comment]
I tried tosee what happens on clean project with  new arch enabled.

### Config:

```
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.74.2
    wanted: 0.74.2
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: true
  ```
  
### Code sample:

```tsx
import React from 'react';
import {Button, SafeAreaView, Text, TextInput} from 'react-native';

const msgs: string[] = [];

function App(): React.JSX.Element {
  const [text, setText] = React.useState('');

  const handlePress = () => {
    msgs.push(text);
    setText('');
  };

  return (
    <SafeAreaView>
      <TextInput value={text} onChangeText={setText} />
      <Button onPress={handlePress} title="Submit" />
      {msgs.map((msg, i) => (
        <Text key={i}>{msg}</Text>
      ))}
    </SafeAreaView>
  );
}

export default App;
```

### Results

https://github.com/Expensify/App/assets/62747088/82c08d49-b96b-405b-9b23-1f5bffaa5f9f

The bug seems to be present in such conditions as well, even without any additional overhead. It looks like an inherent issue with RN itself (please correct me if I am wrong). I will continue diving into how we can manage this, as it makes UX quite unpleasant especially with chat-like app.

**Comment by muttmuure** [comment]
@kacper-mikolajczak could you post these updates in the channel canvas? https://expensify.slack.com/archives/C05R2V5GM1S/p1718186137117569

**Comment by kacper-mikolajczak** [comment]
I am working on this in parallel to https://github.com/Expensify/App/issues/42200, so the updates may be delayed, but the general take away is that the issue is coming from RN architecture itself and it is not easy to fix it entirely. 

In order to mitigate the issue as much as possible we would need to offload the amount of calculations that are happening during typing and sending messages.

**Comment by muttmuure** [comment]
@hannojg maybe this is one you can fix?

**Comment by hannojg** [comment]
Sure, I am available!

**Comment by hannojg** [comment]
**Update:**

We have identified what's architecturally causing the issue in fact native / the text input, and have a few solutions in mind (which we will test on Monday).

### The problem

The problem on its full-scale is exemplified in this chart (for android), and can be explained as this:

1. JS Thread: The user calls "clear text input" (e.g. pressing send button)
2. UI Thread: The user continues to type to the text input 
3. UI Thread: A onChangedText update is created for dispatching, which incremented an internal event count
4. UI Thread: The event from JS is received to clear the text, however its event count is outdated, thus the event is ignored
5. JS Thread: receives the text update from the text input with the full text
6. Js Thread: the state is set to the full text + the text input component gets re-rendered with that value
-> Leading to the bug where the text input appears to be never cleared.

```mermaid
sequenceDiagram
  participant RET as ReactEditText (UI)
  participant RTIM as ReactTextInputManager (UI)
  participant SW as StateWrapper (UI) [Fabric]
  participant X as UI <> JS
  participant TI as TextInput (JS)

  RTIM ->> RET: addTextChangedListener

  Note over TI: User clears Text
  TI ->> TI: React.setState("")
  activate TI

  Note over RET: User enters another "a"
  RET ->> RTIM: onTextChanged
  RTIM ->> SW: updateState
  RTIM ->> X: dispatchEvent ReactTextChangedEvent

  TI ->> X: react update value prop / setText ""
  deactivate TI

  X ->> TI: onChangeText "aa"
  Note over TI: Updates mostRecentEventCount
  TI ->> TI: Updates state rerenders with new value prop "aa"
  TI ->> X: Update props/state (with recent event count)

  X ->> RTIM: receiveCommand / maybeSetTextFromJS ""
  Note over RTIM: Ignore JS update / not in sync
    X ->> SW: updateState (with recent event count)
  SW ->> RTIM: updateExtraData "aa"
  RTIM ->> RET: maybeSetTextFromState

%% Two changes to fix this:
%%  1. Always allow clear dispite event mismatch
%%  2. Only apply partial updates from state / onTextChange
```

It's basically a synchronisation problem between two threads. This is a common problem in software development and there are some patterns for solving it.

### Solutions

**Having a single source of truth and lock:** This can either be on the JS thread or the UI thread

- It's not really a solution to be on the JS thread. This would cause the UI text input to be laggy when the JS thread performs < 60 FPS, which is not acceptable
- One solution we want to investigate is to do everything from the UI thread. This would require a bit of a rewrite of the Composer for all parts that currently rely on the JS `value`.

**Implement a specific mechanism for forcing an update (such as clearing)** 

When a "clear" event is received on the UI thread we can either:

- block it temporarily until the clearing has been completed. The downside here would be that the UI thread / text input is blocked for that short duration. We have to test whether this will be noticeable.
- Implement internally to only apply the new difference of the input text to the input, and always allow for clearing it (meaning to ignore the most recent event count here). This way it would work like:
	- text input is "a"
	- User presses clear
	- User enters next letter "a"
	- Text input is cleared
	- New native update event won't be "aa" as before, but only the text difference to before, which is just "a", so the new text would be "a"

**tl;dr** We will test a few different solutions next week, checking what works well and what doesn't. One solution being to write a custom text input component that entirely operates on the UI thread.

**Comment by hannojg** [comment]
**Update:** Works equally good on iOS:

| Before | ✨ After ✨ |
|--------|--------|
| <video src="https://github.com/Expensify/App/assets/16821682/10e25b2a-3040-4d73-bd89-e297468b3881" /> | <video src="https://github.com/Expensify/App/assets/16821682/53223b53-0270-4d35-bb67-d18c068b7588" /> |

This being said I want to point out that if the JS thread is very busy we can still run into states where the text input isn't in perfect sync with what the user would expect.

Thats because the event to update the text after the clearance has happened is send late after the user has pressed the send/clear button, as the JS thread is so busy performing other tasks first.
In the expensify implementation we already helped and implemented the clearance of text on the UI thread, which should help, so the delay shouldn't be as noticeable as in the video shown above:

https://github.com/Expensify/App/blob/0b4812d10c8a5c6dcc0a0d042b4188c0ffd94979/src/pages/home/report/ReportActionCompose/ReportActionCompose.tsx#L368

**Comment by hannojg** [comment]
This is a first PR where I cleanup some code:

 - https://github.com/Expensify/App/pull/44369

**Comment by hannojg** [comment]
**Update:** I finished the main implementation/refactoring which was to migrate from `onChangeText` to `onChange` and only apply diff updates to our state.

There are still some bugs / edge cases I need to address. In the video you can see that the composer will be cleared once the JS thread catches up. However all text entered after the send gets currently dropped.

Also I feel like there is another major performance issue somewhere around the composer code. While working on it I have seen multiple warnings about "Maximum update depth: Excessive state updates". So I am pretty sure there is something that can be fixed, which maybe makes our composer fast again so the text can be cleared immediately.
(I observed that when I click on send but continue typing the message is never send. I assume thats because typing completely kills the JS thread performance [for some reason I am about to find out]).

I also need to add web/desktop support.
New ETA: tmrw I should have a fully testable PR!


https://github.com/Expensify/App/assets/16821682/58ee5b80-1717-4652-b970-fc0ff7d3a48b

**Comment by hannojg** [comment]
**Update:** I now fixed a few bugs and added web support https://github.com/Expensify/App/pull/44285

However, I just realised there might be a second way to implement the diff text updates using the selection value, which might be simpler / requires less changes in react native.
Will evaluate tmrw and finish the PR tmrw!

**Comment by hannojg** [comment]
**Update:** Today i had limited availability. I did some more testing on my PR and realised that the web part needs a bit more work. Will finish it off early next week!

**Comment by hannojg** [comment]
**Update:** Worked on the web bugs, should be all good now except [for one bug I am still working on](https://github.com/Expensify/App/pull/44285#issuecomment-2200524436).

However, except for this one bug everything seems to work for me.
I will fix the bug tomorrow, meanwhile it would be super awesome if someone could already test the PR!

**Comment by kevinksullivan** [comment]
Assuming this should be DAILY if it's critical? Lmk if I am misunderstanding!

**Comment by hannojg** [comment]
Update: We are waiting on the following PR to be merged first, as it contains the changes I need in my PR to fix this bug:

- https://github.com/Expensify/App/pull/44958

**Comment by hannojg** [comment]
Updated with main which includes the latest rn-live-markdown version with the fixes. However there are a few bugs on web now. Fixing them today!

**Comment by hannojg** [comment]
Opened two PRs in live-markdown to address these bugs

**Comment by hannojg** [comment]
Possible, will take a look tmrw!

**Comment by hannojg** [comment]
The good part is though that I can't reproduce the original issue - same for you @Ollyws ?

**Comment by hannojg** [comment]
The fix has actually been reverted due to a regression!

I am working on a new PR here:

- https://github.com/Expensify/App/pull/46091

**Comment by hannojg** [comment]
[The last PR](https://github.com/Expensify/App/pull/46091) for fixing this was reverted. Working on a new one to fix the regression!

**Comment by hannojg** [comment]
Opened a new PR here:

- https://github.com/Expensify/App/pull/46796

**Comment by muttmuure** [comment]
$250 for @Ollyws - C+

**Comment by muttmuure** [comment]
Actually how many PRs did you review? It seems like this should be more than $250

**Comment by muttmuure** [comment]
Handling

**Comment by muttmuure** [comment]
$750 for @Ollyws C+

**Comment by garrettmknight** [comment]
$750 approved for @Ollyws based on [this comment](https://github.com/Expensify/App/issues/37896#issuecomment-2321935695)
